modules/xfem/src/userobjects/InterfaceMeshCutUserObjectBase.C

```diff
@@ -0,0 +1,112 @@
+  return params;
+}
+
+InterfaceMeshCutUserObjectBase::InterfaceMeshCutUserObjectBase(const InputParameters & parameters)
+  : GeometricCutUserObject(parameters),
+    _interface_velocity(nullptr),
+    _func(nullptr),
+    _mesh(_subproblem.mesh()),
+    _exodus_io(nullptr),
+    _explicit_system(nullptr),
+    _equation_systems(nullptr),
+    _output_exodus(getParam<bool>("output_exodus")),
+    _negative_id(getParam<CutSubdomainID>("negative_id")),
+    _positive_id(getParam<CutSubdomainID>("positive_id"))
+{
+  MeshFileName xfem_cutter_mesh_file = getParam<MeshFileName>("mesh_file");
+  _cutter_mesh = std::make_shared<ReplicatedMesh>(_communicator);
+  _cutter_mesh->read(xfem_cutter_mesh_file);
+  _exodus_io = std::make_unique<ExodusII_IO>(*_cutter_mesh);
+
+  if (!parameters.isParamSetByUser("interface_velocity_uo") &&
+      !parameters.isParamSetByUser("interface_velocity_function"))
+    paramError("Either `interface_velocity_uo` or `interface_velocity_function` must be provided.");
+  else if (parameters.isParamSetByUser("interface_velocity_uo") &&
+           parameters.isParamSetByUser("interface_velocity_function"))
+    paramError(
+        "'interface_velocity_uo` and `interface_velocity_function` cannot be both provided.");
+
+  if (parameters.isParamSetByUser("interface_velocity_function"))
+    _func = &getFunction("interface_velocity_function");
+}
+
+void
+InterfaceMeshCutUserObjectBase::initialSetup()
+{
+  if (_func == nullptr)
+  {
+    const UserObject * uo =
+        &(_fe_problem.getUserObjectBase(getParam<UserObjectName>("interface_velocity_uo")));
+
+    if (dynamic_cast<const XFEMMovingInterfaceVelocityBase *>(uo) == nullptr)
+      mooseError("UserObject casting to XFEMMovingInterfaceVelocityBase in "
+                 "MovingLineSegmentCutSetUserObject");
+
+    _interface_velocity = dynamic_cast<const XFEMMovingInterfaceVelocityBase *>(uo);
+    const_cast<XFEMMovingInterfaceVelocityBase *>(_interface_velocity)->initialize();
+  }
+
+  for (const auto & node : _cutter_mesh->node_ptr_range())
+    _initial_nodes_location[node->id()] = *node;
+
+  for (const auto & elem : _cutter_mesh->element_ptr_range())
+    for (unsigned int n = 0; n < elem->n_nodes(); n++)
+      _node_to_elem_map[elem->node_id(n)].push_back(elem->id());
+
+  _cutter_mesh->prepare_for_use();
+  _cutter_mesh->set_mesh_dimension(_mesh.dimension() - 1);
+
+  if (_output_exodus)
+  {
+    _equation_systems = libmesh_make_unique<EquationSystems>(*_cutter_mesh);
+    _explicit_system = &(_equation_systems->add_system<ExplicitSystem>("InterfaceMeshSystem"));
+
+    _explicit_system->add_variable("disp_x");
+    _explicit_system->add_variable("disp_y");
+
+    if (_mesh.dimension() == 3)
+      _explicit_system->add_variable("disp_z");
+
+    _equation_systems->init();
+    _exodus_io->write_equation_systems(_app.getOutputFileBase() + "_" + name() + ".e",
+                                       *_equation_systems);
+  }
```

Comment:
Could this exodus I/O be a separate member function so that derived classes can reuse it?